### PR TITLE
Be stricter about JSON that is accepted by Synapse

### PR DESCRIPTION
**Pull request by clokep**
This disables the JSON extensions which Python supports by default (parsing of `Infinity` / `-Infinity` and `NaN`). These shouldn't be accepted since they're not technically valid JSON and other languages won't be able to interpret it properly.

I also restructured the JSON deserializing a bit to refer to a single JSON decoder with a shared implementation (kind of like #7372).

It is possible this could break due to event data in someone's database, but that event isn't actually JSON, so I think this is equivalent to somehow ending up with total garbage in the database. Synapse should error pretty clearly in this case anyway.

### CODE REVIEW

File: synapse/util/__init__.py [hunk 1]
```diff
@@ -25,9 +25,18 @@
...
 logger = logging.getLogger(__name__)
 
+
+def _reject_invalid_json(val):
+    """Do not allow Infinity, -Infinity, or NaN values in JSON."""
+    raise json.JSONDecodeError("Invalid JSON value: '%s'" % val)
+
+
 # Create a custom encoder to reduce the whitespace produced by JSON encoding.
 json_encoder = json.JSONEncoder(separators=(",", ":"))
```
Comment (erikjohnston) [inline synapse/util/__init__.py:35]: Can we also do the same here and set `allow_nan=False` to stop us from sending invalid json too?

Comment (clokep) [reply synapse/util/__init__.py:35]: Good call! 👍

Comment (clokep) [reply synapse/util/__init__.py]: Done in 248d8284fa850289689f9ae87d2c807b58d7a812.

There's a few places we don't use this encoder, I was going to move those over in a follow-up (but I can do it here if you'd like).

Comment (clokep) [reply synapse/util/__init__.py]: (Hmm...we should probably make this change to canonicaljson too. I'll file an issue there.)

Comment (clokep) [reply synapse/util/__init__.py]: See matrix-org/python-canonicaljson#32.

**Comment by clokep** [comment]
> Can you check the code for fetching events from the DB to see how they handle json parse errors? Since we fetch events in batches I have a suspicion that it will fail the entire batch, which may or may not be the right thing to do

I looked at this a bit and you're right -- it would fail the whole batch. I pushed 5516ae216d8cfe34b04ad190d8ec3c50bec07835 which will end up treating the event as if it is redacted / broken in some other way (other examples in that code of "broken" could mean the room version event format doesn't match the event's format). This seems consistent with our other error handling.

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**